components/ws-daemon/pkg/cgroup/plugin_iolimit_v2.go

```diff
@@ -6,138 +6,140 @@ package cgroup
 
 import (
 	"context"
-	"fmt"
-	"os"
 	"path/filepath"
+	"strconv"
 	"strings"
-	"time"
+	"sync"
 
+	v2 "github.com/containerd/cgroups/v2"
 	"github.com/gitpod-io/gitpod/common-go/log"
 )
 
-var clearLimits = ioLimitOptions{
-	WriteBytesPerSecond: 0,
-	ReadBytesPerSecond:  0,
-	WriteIOPs:           0,
-	ReadIOPs:            0,
-}
+type IOLimiterV2 struct {
+	limits *v2.Resources
 
-type ioLimitOptions struct {
-	WriteBytesPerSecond int64
-	ReadBytesPerSecond  int64
-	WriteIOPs           int64
-	ReadIOPs            int64
-}
+	cond *sync.Cond
 
-type IOLimiterV2 struct {
-	limits ioLimitOptions
+	devices []string
 }
 
-func NewIOLimiterV2(writeBytesPerSecond, readBytesPerSecond, writeIOPs, readIOPs int64) *IOLimiterV2 {
-	limits := ioLimitOptions{
-		WriteBytesPerSecond: writeBytesPerSecond,
-		ReadBytesPerSecond:  readBytesPerSecond,
-		WriteIOPs:           writeIOPs,
-		ReadIOPs:            readIOPs,
-	}
-
+func NewIOLimiterV2(writeBytesPerSecond, readBytesPerSecond, writeIOPs, readIOPs int64) (*IOLimiterV2, error) {
+	devices := buildDevices()
+	log.WithField("devices", devices).Info("io limiting devices")
 	return &IOLimiterV2{
-		limits: limits,
-	}
+		limits: buildV2Limits(writeBytesPerSecond, readBytesPerSecond, writeIOPs, readIOPs, devices),
+
+		cond:    sync.NewCond(&sync.Mutex{}),
+		devices: devices,
+	}, nil
 }
 
 func (c *IOLimiterV2) Name() string  { return "iolimiter-v2" }
 func (c *IOLimiterV2) Type() Version { return Version2 }
 
 func (c *IOLimiterV2) Apply(ctx context.Context, basePath, cgroupPath string) error {
+	update := make(chan struct{}, 1)
+	go func() {
+		// TODO(cw): this Go-routine will leak per workspace, until we update config or restart ws-daemon
+		defer close(update)
+
+		for {
+			c.cond.L.Lock()
+			c.cond.Wait()
+			c.cond.L.Unlock()
+
+			if ctx.Err() != nil {
+				return
+			}
+
+			update <- struct{}{}
+		}
+	}()
+
 	go func() {
-		log.WithField("cgroupPath", cgroupPath).Debug("starting io limiting")
-		// We are racing workspacekit and the interaction with disks.
-		// If we did this just once there's a chance we haven't interacted with all
-		// devices yet, and hence would not impose IO limits on them.
-		ticker := time.NewTicker(30 * time.Second)
-		defer ticker.Stop()
+		log.WithField("cgroupPath", cgroupPath).Info("starting io limiting")
 
-		ioMaxFile := filepath.Join(basePath, cgroupPath)
+		_, err := v2.NewManager(basePath, filepath.Join("/", cgroupPath), c.limits)
```

Comment:
We only want to write the value to `io.max` , right? Why do you use contained Manager?

Reply:
It was a little difficult to know what this code means because I had to assume some kinds of things.

Reply:
I actually mistakenly thought this was updating the device filter and changing the eBPF. However, I admit I was a little dumb.